### PR TITLE
spack concretize: use glyphs for spec status

### PR DESCRIPTION
**Pull request by Paul-Ferrell**
I regularly use the `spack concretize` command to get a sense of what `spack install` will do in environment before I actually run it. This is especially important when using upstreams, as you really want to make sure you're using upstream packages. I felt the command needed some improvement.

This merge request:
 - Adds documentation for what the 'package status' glyphs mean. 
 - Changes the glyphs to be a bit more obvious
   - All are in square brackets `[-]` and bold.
   - Installed is now `[✔]`
   - Missing is now `[?]`
   - Upstream is now blue.
 - Updated the corresponding glyph that's written when installing to a checkmark.
 - Install status now also checks that the install location exists.
 - Added a `-d/--dry-run` option to `spack concretize`, to see the tree without
   actually updating the spack.lock file.

Screenshot from 2020-05-06 12-05-25

### CODE REVIEW

File: lib/spack/spack/cmd/concretize.py [hunk 1]
```diff
@@ -4,8 +4,16 @@
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
 import spack.environment as ev
+import llnl.util.tty.color as clr
 
-description = 'concretize an environment and write a lockfile'
+description = clr.colorize(
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/concretize.py]: The description is supposed to be a short one-line summary. I would add the extra info you want to `subparser.epilog` so it gets displayed in the help message. Note that `spack spec -I` uses the same glyphs, so you may want to add documentation there too.

**Review by adamjstewart — COMMENTED** [review]
I think the glyph changes may be a bit controversial, but the rest of the PR should be fine

File: lib/spack/spack/cmd/concretize.py (hunk 1 above)
Comment (Paul-Ferrell) [reply lib/spack/spack/cmd/concretize.py]: Install and buildcache commands use it too, although build cache uses a different `status_fn` so the meanings are totally different. 
I moved the glyph definitions to class attributes so they can be used directly in the help output, and there's no continuity issues if they ever change again.
I also documented what sort of function is expected for the `status_fn`, and changed Spec.install_status to be consistent with that. There's only one instance of a different `status_fn` being used, and it only kind-of happens to work. 

It's a bit broken that the 'upstream' status is calculated independent of that function, but for the moment that doesn't matter.

File: lib/spack/spack/cmd/buildcache.py [hunk 2]
```diff
@@ -129,8 +129,8 @@ def setup_parser(subparser):
 
     preview_parser = subparsers.add_parser(
         'preview',
-        help='analyzes an installed spec and reports whether '
-             'executables and libraries are relocatable'
+        help='Analyzes an installed spec and reports whether '
+             'executables and libraries are relocatable.'
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/buildcache.py]: I would keep these help messages as lowercase without a period, that's the default argparse uses for things like `--help`

**Comment by tgamblin** [comment]
This LGTM but do the glyphs show up reliably on different platforms?  Don't we need to check the system locale for this to see whether the glyphs are in the supported charset?

Also, the `[+]` is used for `spack install` as well.  If we change it to `[✔]` in `spack spec`, shouldn't we also change it there?

**Comment by Paul-Ferrell** [comment]
@tgamblin - I thought about font issues, and checked Mac and Linux, but forgot about Windows. Only one of the fonts listed under PowerShell on Windows 10 supports any unicode, and that one is broken in other ways. 

So, I guess a check isn't going to happen, unless we do things differently on Windows than other platforms. (Which is disappointing, the check is immediately obvious in meaning). 

As for the `spack install` `[+]` it was already changed in this commit too.

**Comment by adamjstewart** [comment]
For the record, Unicode glyphs/emoji don't currently work on WSL: https://github.com/microsoft/terminal/issues/190

They do seem to work in the newer Windows Terminal though. Haven't tried Command Prompt, PowerShell, Cygwin, etc. etc. etc.

Really wish we wouldn't have to support the 5+ different terminals that Windows provides...

**Comment by tldahlgren** [comment]
Should external packages be flagged separately as is being done for upstream?  I can imagine this could be useful to know as well.

**Comment by alalazo** [comment]
What is the status of this PR?

**Comment by Paul-Ferrell** [comment]
I need to drop the glyphs, they're too OS specific, and update this to merge properly with the latest Spack.
I've got a hot minute, now's a good time as any...

**Comment by alalazo** [comment]
Thanks for the prompt update @Paul-Ferrell

**Comment by Paul-Ferrell** [comment]
Closing

**Comment by Paul-Ferrell** [comment]
Closing this and replacing with #24280. (There were issues with the original branch).

**Pull request closed without merging**